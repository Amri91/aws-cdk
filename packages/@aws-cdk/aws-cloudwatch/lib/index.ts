export * from './alarm';
export * from './alarm-action';
export * from './dashboard';
export * from './graph';
export * from './layout';
export * from './metric';
export * from './metric-types';
export * from './text';
export * from './widget';
export * from './timeseries';
export * from './expression';

// AWS::CloudWatch CloudFormation Resources:
export * from './cloudwatch.generated';
