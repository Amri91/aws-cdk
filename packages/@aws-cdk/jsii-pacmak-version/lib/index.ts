export * from './jsii-pacmak.generated';
