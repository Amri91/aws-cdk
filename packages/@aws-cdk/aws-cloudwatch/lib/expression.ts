import {CfnAlarm} from "./cloudwatch.generated";
import {AlarmTimeSeriesProps, ITimeSeries, TimeSeriesJson, ToJsonProps} from "./timeseries";
import MetricDataQueryProperty = CfnAlarm.MetricDataQueryProperty;

/**
 * TODO
 */
export interface IExpression extends ITimeSeries {
    /**
     * Math or search expression
     */
    expression: string
    /**
     * Unique ID
     */
    id: string
    /**
     * Label for this expression when added to a Graph in a Dashboard
     */
    label?: string
}

/**
 * TODO
 */
export interface ExpressionProps {
    /**
     * TODO
     */
    readonly expression: string
    /**
     * TODO
     */
    readonly id: string
    /**
     * TODO
     * @default TODO
     */
    readonly label?: string
}

/**
 * TODOs
 */
export class Expression implements IExpression {
    /**
     * TODO
     */
    public expression: string;
    /**
     * TODO
     */
    public id: string;
    /**
     * TODO
     * @default TODO
     */
    public label?: string;

    constructor(props: ExpressionProps) {
        this.expression = props.expression;
        this.id = props.id;
        this.label = props.label;
    }

    public toJson(_props: ToJsonProps = {}): TimeSeriesJson {
        return [this];
    }

    public toAlarmTimeSeries(props: AlarmTimeSeriesProps = {}): MetricDataQueryProperty {
        return {
            ...this,
            returnData: props.returnData
        };
    }
}