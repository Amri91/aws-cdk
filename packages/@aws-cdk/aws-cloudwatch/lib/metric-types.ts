import { ITimeSeries } from "./timeseries";

/**
 * Interface for metrics
 */
export interface IMetric extends ITimeSeries {
  /**
   * Turn this metric object into an alarm configuration
   */
  toAlarmConfig(): MetricAlarmConfig;
}

/**
 * Metric dimension
 */
export interface Dimension {
  /**
   * Name of the dimension
   */
  readonly name: string;

  /**
   * Value of the dimension
   */
  readonly value: any;
}

/**
 * Statistic to use over the aggregation period
 */
export enum Statistic {
  SAMPLE_COUNT = 'SampleCount',
  AVERAGE = 'Average',
  SUM = 'Sum',
  MINIMUM = 'Minimum',
  MAXIMUM = 'Maximum',
}

/**
 * Unit for metric
 */
export enum Unit {
  SECONDS = 'Seconds',
  MICROSECONDS = 'Microseconds',
  MILLISECONDS = 'Milliseconds',
  BYTES = 'Bytes',
  KILOBYTES = 'Kilobytes',
  MEGABYTES = 'Megabytes',
  GIGABYTES = 'Gigabytes',
  TERABYTES = 'Terabytes',
  BITS = 'Bits',
  KILOBITS = 'Kilobits',
  MEGABITS = 'Megabits',
  GIGABITS = 'Gigabits',
  TERABITS = 'Terabits',
  PERCENT = 'Percent',
  COUNT = 'Count',
  BYTES_PER_SECOND = 'Bytes/Second',
  KILOBYTES_PER_SECOND = 'Kilobytes/Second',
  MEGABYTES_PER_SECOND = 'Megabytes/Second',
  GIGABYTES_PER_SECOND = 'Gigabytes/Second',
  TERABYTES_PER_SECOND = 'Terabytes/Second',
  BITS_PER_SECOND = 'Bits/Second',
  KILOBITS_PER_SECOND = 'Kilobits/Second',
  MEGABITS_PER_SECOND = 'Megabits/Second',
  GIGABITS_PER_SECOND = 'Gigabits/Second',
  TERABITS_PER_SECOND = 'Terabits/Second',
  COUNT_PER_SECOND = 'Count/Second',
  NONE = 'None'
}

/**
 * Properties used to construct the Metric identifying part of an Alarm
 */
export interface MetricAlarmConfig {
  /**
   * The dimensions to apply to the alarm
   */
  readonly dimensions?: Dimension[];

  /**
   * Namespace of the metric
   */
  readonly namespace: string;

  /**
   * Name of the metric
   */
  readonly metricName: string;

  /**
   * How many seconds to aggregate over
   */
  readonly period: number;

  /**
   * Simple aggregation function to use
   */
  readonly statistic?: Statistic;

  /**
   * Percentile aggregation function to use
   */
  readonly extendedStatistic?: string;

  /**
   * The unit of the alarm
   */
  readonly unit?: Unit;
}

/**
 * TODO
 */
export type yAxisType = 'left' | 'right';

/**
 * Custom rendering properties that override the default rendering properties specified in the yAxis parameter of the widget object.
 */
export interface MetricRenderingProperties {
  /**
   * How many seconds to aggregate over
   */
  readonly period: number;

  /**
   * Label for the metric
   */
  readonly label?: string;

  /**
   * Color for the graph line
   */
  readonly color?: string;

  /**
   * Aggregation function to use (can be either simple or a percentile)
   */
  readonly stat?: string;

  /**
   * TODO
   * @default TODO
   */
  readonly visible?: boolean,

  /**
   * TODO
   * @default "left"
   */
  readonly yAxis: yAxisType

  /**
   * TODO
   * @default todo
   */
  readonly id?: string
}