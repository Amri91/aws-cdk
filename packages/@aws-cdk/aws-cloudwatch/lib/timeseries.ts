import { CfnAlarm } from "./cloudwatch.generated";
import { IExpression } from "./expression";
import { yAxisType } from "./metric-types";
import MetricDataQueryProperty = CfnAlarm.MetricDataQueryProperty;

export type MetricJson = [string, string, ...any[]];
export type ExpressionJson = [IExpression];
export type TimeSeriesJson = ExpressionJson|MetricJson;

/**
 * TODO
 */
export interface ToJsonProps {
    /**
     * TODO
     * @default TODO
     */
    readonly yAxis?: yAxisType,

    /**
     * TODO
     * @default TODO
     */
    readonly visible?: boolean
}

/**
 * TODO
 */
export interface AlarmTimeSeriesProps {
    /**
     * TODO
     * @default TODO
     */
    readonly returnData?: boolean
}

/**
 * TODO
 */
export interface ITimeSeries {
    // TODO: should add properties?

    /**
     * Returns an the array representation of this timeseries.
     */
    toJson(props?: ToJsonProps): TimeSeriesJson
    /**
     * Returns a configuration entry suitable for alarms.Metrics
     */
    toAlarmTimeSeries(props?: AlarmTimeSeriesProps): MetricDataQueryProperty
}