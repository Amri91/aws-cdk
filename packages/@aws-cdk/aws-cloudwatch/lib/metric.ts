import iam = require('@aws-cdk/aws-iam');
import cdk = require('@aws-cdk/core');
import { ComparisonOperator, TreatMissingData } from './alarm';
import { CfnAlarm } from './cloudwatch.generated';
import { Dimension, IMetric, MetricAlarmConfig, MetricRenderingProperties, Unit } from './metric-types';
import { AlarmTimeSeriesProps, MetricJson, TimeSeriesJson, ToJsonProps } from './timeseries';
import { normalizeStatistic, parseStatistic } from './util.statistic';

export type DimensionHash = {[dim: string]: any};

/**
 * Options shared by most methods accepting metric options
 */
export interface CommonMetricOptions {
  /**
   * The period over which the specified statistic is applied.
   *
   * @default Duration.minutes(5)
   */
  readonly period?: cdk.Duration;

  /**
   * What function to use for aggregating.
   *
   * Can be one of the following:
   *
   * - "Minimum" | "min"
   * - "Maximum" | "max"
   * - "Average" | "avg"
   * - "Sum" | "sum"
   * - "SampleCount | "n"
   * - "pNN.NN"
   *
   * @default Average
   */
  readonly statistic?: string;

  /**
   * Dimensions of the metric
   *
   * @default - No dimensions.
   */
  readonly dimensions?: DimensionHash;

  /**
   * Unit for the metric that is associated with the alarm
   */
  readonly unit?: Unit;

  /**
   * Label for this metric when added to a Graph in a Dashboard
   */
  readonly label?: string;

  /**
   * Color for this metric when added to a Graph in a Dashboard
   */
  readonly color?: string;

  /**
   * TODO
   * @default todo
   */
  readonly id?: string;
}

/**
 * Properties for a metric
 */
export interface MetricProps extends CommonMetricOptions {
  /**
   * Namespace of the metric.
   */
  readonly namespace: string;

  /**
   * Name of the metric.
   */
  readonly metricName: string;
}

/**
 * Properties of a metric that can be changed
 */
export interface MetricOptions extends CommonMetricOptions {
}

/**
 * A metric emitted by a service
 *
 * The metric is a combination of a metric identifier (namespace, name and dimensions)
 * and an aggregation function (statistic, period and unit).
 *
 * It also contains metadata which is used only in graphs, such as color and label.
 * It makes sense to embed this in here, so that compound constructs can attach
 * that metadata to metrics they expose.
 *
 * This class does not represent a resource, so hence is not a construct. Instead,
 * Metric is an abstraction that makes it easy to specify metrics for use in both
 * alarms and graphs.
 */
export class Metric implements IMetric {
  /**
   * Grant permissions to the given identity to write metrics.
   *
   * @param grantee The IAM identity to give permissions to.
   */
  public static grantPutMetricData(grantee: iam.IGrantable): iam.Grant {
    return iam.Grant.addToPrincipal({
      grantee,
      actions: ['cloudwatch:PutMetricData'],
      resourceArns: ['*']
    });
  }

  public readonly dimensions?: DimensionHash;
  public readonly namespace: string;
  public readonly metricName: string;
  public readonly period: cdk.Duration;
  public readonly statistic: string;
  public readonly unit?: Unit;
  public readonly label?: string;
  public readonly color?: string;
  /**
   * TODO
   */
  public readonly id: string;

  constructor(props: MetricProps) {
    this.period = props.period || cdk.Duration.minutes(5);
    const periodSec = this.period.toSeconds();
    if (periodSec !== 1 && periodSec !== 5 && periodSec !== 10 && periodSec !== 30 && periodSec % 60 !== 0) {
      throw new Error(`'period' must be 1, 5, 10, 30, or a multiple of 60 seconds, received ${props.period}`);
    }

    this.dimensions = props.dimensions;
    this.namespace = props.namespace;
    this.metricName = props.metricName;
    // Try parsing, this will throw if it's not a valid stat
    this.statistic = normalizeStatistic(props.statistic || "Average");
    this.label = props.label;
    this.color = props.color;
    this.unit = props.unit;
    this.id = props.id || this.metricName.toLowerCase();
  }

  public toAlarmTimeSeries(props: AlarmTimeSeriesProps = {}): CfnAlarm.MetricDataQueryProperty {
    return {
      metricStat: {
        metric: {
          dimensions: this.dimensionsAsList(),
          metricName: this.metricName,
          namespace: this.namespace,
        },
        period: this.period.toSeconds(),
        stat: this.statistic,
        unit: this.unit
      },
      id: this.id,
      returnData: props.returnData
    };
  }

  public toJson(props: ToJsonProps = {}): TimeSeriesJson {
    const dimensions: any[] = this.dimensionsAsList().map(d => [d.name, d.value]);
    const dims = [].concat(...dimensions);

    const renderingProperties: MetricRenderingProperties = {
      color: this.color,
      label: this.label,
      period: this.period.toSeconds(),
      stat: this.statistic,
      visible: props.visible,
      yAxis: props.yAxis || "left",
      id: this.id
    };

    // Todo: refactor
    const array: MetricJson = [
      this.namespace,
      this.metricName,
      ...dims,
      // renderingProperties
    ];
    array.push(renderingProperties);

    return array;
  }

  public toAlarmConfig(): MetricAlarmConfig {
    const stat = parseStatistic(this.statistic);
    const dims = this.dimensionsAsList();

    return {
      dimensions: dims.length > 0 ? dims : undefined,
      namespace: this.namespace,
      metricName: this.metricName,
      period: this.period.toSeconds(),
      statistic: stat.type === 'simple' ? stat.statistic : undefined,
      extendedStatistic: stat.type === 'percentile' ? 'p' + stat.percentile : undefined,
      unit: this.unit
    };
  }

  public toString() {
    return this.label || this.metricName;
  }

  /**
   * Return the dimensions of this Metric as a list of Dimension.
   */
  private dimensionsAsList(): Dimension[] {
    const dims = this.dimensions;

    if (dims === undefined) {
      return [];
    }

    return Object.keys(dims).map(key => ({ name: key, value: dims[key] }));
  }
}

/**
 * TODO
 */
export interface BaseAlarmProps {
  /**
   * Name of the alarm
   *
   * @default Automatically generated name
   */
  readonly alarmName?: string;

  /**
   * Description for the alarm
   *
   * @default No description
   */
  readonly alarmDescription?: string;

  /**
   * Sets how this alarm is to handle missing data points.
   *
   * @default TreatMissingData.Missing
   */
  readonly treatMissingData?: TreatMissingData;

  /**
   * Comparison to use to check if metric is breaching
   *
   * @default GreaterThanOrEqualToThreshold
   */
  readonly comparisonOperator?: ComparisonOperator;

  /**
   * The value against which the specified statistic is compared.
   */
  readonly threshold: number;

  /**
   * The number of periods over which data is compared to the specified threshold.
   */
  readonly evaluationPeriods: number;

  /**
   * Specifies whether to evaluate the data and potentially change the alarm state if there are too few data points to be statistically significant.
   *
   * Used only for alarms that are based on percentiles.
   *
   * @default - Not configured.
   */
  readonly evaluateLowSampleCountPercentile?: string;

  /**
   * Whether the actions for this alarm are enabled
   *
   * @default true
   */
  readonly actionsEnabled?: boolean;

  /**
   * The number of datapoints that must be breaching to trigger the alarm. This is used only if you are setting an "M
   * out of N" alarm. In that case, this value is the M. For more information, see Evaluating an Alarm in the Amazon
   * CloudWatch User Guide.
   *
   * @default ``evaluationPeriods``
   *
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html#alarm-evaluation
   */
  readonly datapointsToAlarm?: number;
}

/**
 * Properties needed to make an alarm from a metric
 */
export interface CreateAlarmOptions extends BaseAlarmProps {
  /**
   * The period over which the specified statistic is applied.
   *
   * @default Duration.minutes(5)
   */
  readonly period?: cdk.Duration;

  /**
   * What function to use for aggregating.
   *
   * Can be one of the following:
   *
   * - "Minimum" | "min"
   * - "Maximum" | "max"
   * - "Average" | "avg"
   * - "Sum" | "sum"
   * - "SampleCount | "n"
   * - "pNN.NN"
   *
   * @default Average
   */
  readonly statistic?: string;
}
