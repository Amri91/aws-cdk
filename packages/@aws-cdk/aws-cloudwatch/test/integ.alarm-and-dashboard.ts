// Integration test to deploy some resources, create an alarm on it and create a dashboard.
//
// Because literally every other library is going to depend on @aws-cdk/aws-cloudwatch, we drop down
// to the very lowest level to create CloudFormation resources by hand, without even generated
// library support.

import cdk = require('@aws-cdk/core');
import cloudwatch = require('../lib');
import { PeriodOverride } from '../lib';

const app = new cdk.App();

const stack = new cdk.Stack(app, `aws-cdk-cloudwatch`);

const queue = new cdk.CfnResource(stack, 'queue', { type: 'AWS::SQS::Queue' });

const metric = new cloudwatch.Metric({
  namespace: 'AWS/SQS',
  metricName: 'ApproximateNumberOfMessagesVisible',
  dimensions: { QueueName: queue.getAtt('QueueName') }
});

const mathExpression = new cloudwatch.Expression({expression: 'SUM(METRICS())', id: 'e1'});

const alarm = new cloudwatch.Alarm(stack, 'Alarm', {
  metric,
  threshold: 100,
  evaluationPeriods: 3,
  datapointsToAlarm: 2
});

const complexAlarm = new cloudwatch.Alarm(stack, 'ComplexAlarm', {
  alarmTimeSeries: [metric, mathExpression].map((ats, index) => ats.toAlarmTimeSeries({returnData: index === 1})),
  threshold: 100,
  evaluationPeriods: 3
});

const dashboard = new cloudwatch.Dashboard(stack, 'Dash', {
  dashboardName: 'MyCustomDashboardName',
  start: '-9H',
  end: '2018-12-17T06:00:00.000Z',
  periodOverride: PeriodOverride.INHERIT
});
dashboard.addWidgets(
  new cloudwatch.TextWidget({ markdown: '# This is my dashboard' }),
  new cloudwatch.TextWidget({ markdown: 'you like?' }),
);
dashboard.addWidgets(new cloudwatch.AlarmWidget({
  title: 'Messages in queue',
  alarm,
}));
dashboard.addWidgets(new cloudwatch.GraphWidget({
  title: 'More messages in queue with alarm annotation',
  timeSeries: [metric.toJson()],
  leftAnnotations: [alarm.toAnnotation()]
}));
dashboard.addWidgets(new cloudwatch.SingleValueWidget({
  title: 'Current messages in queue',
  timeSeries: [metric.toJson()]
}));

dashboard.addWidgets(new cloudwatch.GraphWidget({
  timeSeries: [metric, mathExpression].map(ts => ts.toJson()),
  title: 'My time series'
}));

dashboard.addWidgets(new cloudwatch.AlarmWidget({
  alarm: complexAlarm
}));

app.synth();
