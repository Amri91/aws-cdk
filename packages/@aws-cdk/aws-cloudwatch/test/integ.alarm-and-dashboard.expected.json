{
  "Resources": {
    "queue": {
      "Type": "AWS::SQS::Queue"
    },
    "Alarm7103F465": {
      "Type": "AWS::CloudWatch::Alarm",
      "Properties": {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 3,
        "DatapointsToAlarm": 2,
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "queue",
                "QueueName"
              ]
            }
          }
        ],
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 100
      }
    },
    "ComplexAlarmA8F1D103": {
      "Type": "AWS::CloudWatch::Alarm",
      "Properties": {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 3,
        "Metrics": [
          {
            "Id": "approximatenumberofmessagesvisible",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "QueueName",
                    "Value": {
                      "Fn::GetAtt": [
                        "queue",
                        "QueueName"
                      ]
                    }
                  }
                ],
                "MetricName": "ApproximateNumberOfMessagesVisible",
                "Namespace": "AWS/SQS"
              },
              "Period": 300,
              "Stat": "Average"
            },
            "ReturnData": false
          },
          {
            "Expression": "SUM(METRICS())",
            "Id": "e1",
            "ReturnData": true
          }
        ],
        "Threshold": 100
      }
    },
    "DashCCD7F836": {
      "Type": "AWS::CloudWatch::Dashboard",
      "Properties": {
        "DashboardBody": {
          "Fn::Join": [
            "",
            [
              "{\"start\":\"-9H\",\"end\":\"2018-12-17T06:00:00.000Z\",\"periodOverride\":\"inherit\",\"widgets\":[{\"type\":\"text\",\"width\":6,\"height\":2,\"x\":0,\"y\":0,\"properties\":{\"markdown\":\"# This is my dashboard\"}},{\"type\":\"text\",\"width\":6,\"height\":2,\"x\":6,\"y\":0,\"properties\":{\"markdown\":\"you like?\"}},{\"type\":\"metric\",\"width\":6,\"height\":6,\"x\":0,\"y\":2,\"properties\":{\"view\":\"timeSeries\",\"title\":\"Messages in queue\",\"region\":\"",
              {
                "Ref": "AWS::Region"
              },
              "\",\"annotations\":{\"alarms\":[\"",
              {
                "Fn::GetAtt": [
                  "Alarm7103F465",
                  "Arn"
                ]
              },
              "\"]},\"yAxis\":{}}},{\"type\":\"metric\",\"width\":6,\"height\":6,\"x\":0,\"y\":8,\"properties\":{\"view\":\"timeSeries\",\"title\":\"More messages in queue with alarm annotation\",\"region\":\"",
              {
                "Ref": "AWS::Region"
              },
              "\",\"metrics\":[[\"AWS/SQS\",\"ApproximateNumberOfMessagesVisible\",\"QueueName\",\"",
              {
                "Fn::GetAtt": [
                  "queue",
                  "QueueName"
                ]
              },
              "\",{\"period\":300,\"stat\":\"Average\",\"yAxis\":\"left\",\"id\":\"approximatenumberofmessagesvisible\"}]],\"annotations\":{\"horizontal\":[{\"label\":\"ApproximateNumberOfMessagesVisible >= 100 for 3 datapoints within 15 minutes\",\"value\":100,\"yAxis\":\"left\"}]},\"yAxis\":{}}},{\"type\":\"metric\",\"width\":6,\"height\":3,\"x\":0,\"y\":14,\"properties\":{\"view\":\"singleValue\",\"title\":\"Current messages in queue\",\"region\":\"",
              {
                "Ref": "AWS::Region"
              },
              "\",\"metrics\":[[\"AWS/SQS\",\"ApproximateNumberOfMessagesVisible\",\"QueueName\",\"",
              {
                "Fn::GetAtt": [
                  "queue",
                  "QueueName"
                ]
              },
              "\",{\"period\":300,\"stat\":\"Average\",\"yAxis\":\"left\",\"id\":\"approximatenumberofmessagesvisible\"}]]}},{\"type\":\"metric\",\"width\":6,\"height\":6,\"x\":0,\"y\":17,\"properties\":{\"view\":\"timeSeries\",\"title\":\"My time series\",\"region\":\"",
              {
                "Ref": "AWS::Region"
              },
              "\",\"metrics\":[[\"AWS/SQS\",\"ApproximateNumberOfMessagesVisible\",\"QueueName\",\"",
              {
                "Fn::GetAtt": [
                  "queue",
                  "QueueName"
                ]
              },
              "\",{\"period\":300,\"stat\":\"Average\",\"yAxis\":\"left\",\"id\":\"approximatenumberofmessagesvisible\"}],[{\"expression\":\"SUM(METRICS())\",\"id\":\"e1\"}]],\"yAxis\":{}}},{\"type\":\"metric\",\"width\":6,\"height\":6,\"x\":0,\"y\":23,\"properties\":{\"view\":\"timeSeries\",\"region\":\"",
              {
                "Ref": "AWS::Region"
              },
              "\",\"annotations\":{\"alarms\":[\"",
              {
                "Fn::GetAtt": [
                  "ComplexAlarmA8F1D103",
                  "Arn"
                ]
              },
              "\"]},\"yAxis\":{}}}]}"
            ]
          ]
        },
        "DashboardName": "MyCustomDashboardName"
      }
    }
  }
}